src/Cafe/OS/libs/nsyshid/Skylander.cpp

```diff
@@ -0,0 +1,400 @@
+		for (auto& s : skylanders)
+		{
+			if (s.status & 1)
+			{
+				s.queued_status.push(3);
+				s.queued_status.push(1);
+			}
+		}
+
+		m_activated = true;
+	}
+
+	void SkylanderUSB::deactivate()
+	{
+		std::lock_guard lock(sky_mutex);
+
+		for (auto& s : skylanders)
+		{
+			// check if at the end of the updates there would be a figure on the portal
+			if (!s.queued_status.empty())
+			{
+				s.status = s.queued_status.back();
+				s.queued_status = std::queue<uint8>();
+			}
+
+			s.status &= 1;
+		}
+
+		m_activated = false;
+	}
+
+	uint8 SkylanderUSB::load_skylander(uint8* buf, std::FILE* file)
+	{
+		std::lock_guard lock(sky_mutex);
+
+		uint32 sky_serial = 0;
+		for (int i = 3; i > -1; i--)
+		{
+			sky_serial <<= 8;
+			sky_serial |= buf[i];
+		}
+		uint8 found_slot = 0xFF;
+
+		// mimics spot retaining on the portal
+		for (auto i = 0; i < 16; i++)
+		{
+			if ((skylanders[i].status & 1) == 0)
+			{
+				if (skylanders[i].last_id == sky_serial)
+				{
+					found_slot = i;
+					break;
+				}
+
+				if (i < found_slot)
+				{
+					found_slot = i;
+				}
+			}
+		}
+
+		if (found_slot != 0xFF)
+		{
+			auto& skylander = skylanders[found_slot];
+			memcpy(skylander.data.data(), buf, skylander.data.size());
+			skylander.sky_file = std::move(file);
+			skylander.status = Skylander::ADDED;
+			skylander.queued_status.push(Skylander::ADDED);
+			skylander.queued_status.push(Skylander::READY);
+			skylander.last_id = sky_serial;
+		}
+		return found_slot;
+	}
+
+	bool SkylanderUSB::remove_skylander(uint8 sky_num)
+	{
+		std::lock_guard lock(sky_mutex);
+		auto& thesky = skylanders[sky_num];
+
+		if (thesky.status & 1)
+		{
+			thesky.status = 2;
+			thesky.queued_status.push(2);
+			thesky.queued_status.push(0);
+			thesky.Save();
+			std::fclose(thesky.sky_file);
+			return true;
+		}
+
+		return false;
+	}
+
+	void SkylanderUSB::query_block(uint8 sky_num, uint8 block, uint8* reply_buf)
+	{
+		std::lock_guard lock(sky_mutex);
+
+		const auto& skylander = skylanders[sky_num];
+
+		reply_buf[0] = 'Q';
+		reply_buf[2] = block;
+		if (skylander.status & 1)
+		{
+			reply_buf[1] = (0x10 | sky_num);
+			memcpy(reply_buf + 3, skylander.data.data() + (16 * block), 16);
+		}
+		else
+		{
+			reply_buf[1] = sky_num;
+		}
+	}
+
+	void SkylanderUSB::write_block(uint8 sky_num, uint8 block,
+								   const uint8* to_write_buf, uint8* reply_buf)
+	{
+		std::lock_guard lock(sky_mutex);
+
+		auto& skylander = skylanders[sky_num];
+
+		reply_buf[0] = 'W';
+		reply_buf[2] = block;
+
+		if (skylander.status & 1)
+		{
+			reply_buf[1] = (0x10 | sky_num);
+			memcpy(skylander.data.data() + (block * 16), to_write_buf, 16);
+			skylander.Save();
+		}
+		else
+		{
+			reply_buf[1] = sky_num;
+		}
+	}
+
+	std::array<uint8, 64> SkylanderUSB::get_status()
+	{
+		std::array<uint8, 64> interrupt_response = {};
+
+		if (!m_queries.empty())
+		{
+			interrupt_response = m_queries.front();
+			m_queries.pop();
```

Comment:
Excuse my paranoia, but I think interacting with `m_queries` like this should be protected by a mutex, and the `m_queries.push` in `SkylanderUSB::control_transfer` on line 197 as well.

Reply:
Yeah a mutex makes sense here